### ISSUE TITLE
Memory leak in long-living effects example

### ISSUE THREAD

**Issue by BeerPoldet** [pointfreeco/swift-composable-architecture#993]
**Describe the bug**
There're memory leaks detected when I have TCA store subscribes to long-running effects which have `cancallable(id:)` attached to. 

**To Reproduce**
Take "Long-living effects" as an example. Go to the screen then tap on "Navigate to another screen", then tap back. The expectation is the long-running effect gets canceled and no memory leak.
<img width="425" alt="Screen Shot 2022-02-05 at 13 21 59" src="https://user-images.githubusercontent.com/7531231/152631343-f64488a1-dae5-4ec4-a44d-1253a767b02c.png">

Below is a screenshot of the leak debugger.
<img width="1574" alt="Screen Shot 2022-02-05 at 13 21 51" src="https://user-images.githubusercontent.com/7531231/152631325-33735e43-b645-4abb-8d94-f5df801b0c9f.png">

Zip up a project that reproduces the behavior and attach it by dragging it here.
Check out the example project from https://github.com/pointfreeco/swift-composable-architecture. 
```swift
let longLivingEffectsReducer = Reducer<
  LongLivingEffectsState, LongLivingEffectsAction, LongLivingEffectsEnvironment
> { state, action, environment in

  struct UserDidTakeScreenshotNotificationId: Hashable {}

  switch action {

  ...

  case .onAppear:
    // When the view appears, start the effect that emits when screenshots are taken.
    return environment.userDidTakeScreenshot
      .map { LongLivingEffectsAction.userDidTakeScreenshotNotification }
      .cancellable(id: UserDidTakeScreenshotNotificationId()) // Without this line, the leak goes away. but cancellation is required to clean subscription up as we know.

  case .onDisappear:
    // When view disappears, stop the effect.
    return .cancel(id: UserDidTakeScreenshotNotificationId())
  }
}

```

**Expected behavior**
`cancellable(id:) should not cause memory leaks.

**Environment**
  - swift-composable-architecture version 0.33.1 (Latest)
  - Xcode 13.2.1
  - Swift 5.5.2
  - OS: iOS 15.2

**Additional context**
Add any more context about the problem here.

**Comment by tgrapperon** [comment]
As I'm getting it, the Effect should be alive iff you're on the "Long-living Effects" screen: `.onDisappear` is called when you navigate to the other screen, and `.onAppear` when you get back to it.
I've also checked that in both cases (navigate to another screen, and pop back to the main list), the effect was canceled and the `cancellationCancellable` was empty.
In other words, I'm not exactly seeing where the effect is not released when it should. Do you see the effect still alive when you're on "another screen"?

**Comment by BeerPoldet** [comment]
The effect was canceled as I expected. But there were memory leaks when I inspected it, It seems to be pointing to (at the image where the line is highlighting)
<img width="1021" alt="Screen Shot 2022-02-05 at 15 04 49" src="https://user-images.githubusercontent.com/7531231/152633908-c87537ff-e0b3-4c1b-965b-d55d08956574.png">

I have no idea if I was tracing to the right location. but at least it's what Xcode is pointing to.

**Comment by tgrapperon** [comment]
This is weird. On Xcode 13.3b1, it's not reporting any leak.

**Comment by tgrapperon** [comment]
I've also experienced false-positive regarding leaks: not all cycles are leaks, as you can sometimes break the cycle manually at appropriate times. This is maybe what's happening here, where we hold onto some values that are released when the subscriptions are explicitly canceled. I would need to investigate further though.

**Comment by tgrapperon** [comment]
To elaborate on my last comment, if the number of leaks doesn't grow with the number of times you go back and forth "another view", these are probably false positives.

**Comment by BeerPoldet** [comment]
Unfortunately, the leaks grow up as I go back and forth 🙁.

**Comment by stephencelis** [comment]
We've encountered a bunch of leaks and bugs in Combine, unfortunately, and there are several APIs in TCA that explicitly work around some of them:

- `Fail` publisher can cause crashes: https://github.com/pointfreeco/swift-composable-architecture/blob/dd084dc147dadede0ad25f829a43ceafed2b9abc/Sources/ComposableArchitecture/Effect.swift#L61-L69
- `Empty(completeImmediately: true)` can never complete: https://github.com/pointfreeco/swift-composable-architecture/blob/dd084dc147dadede0ad25f829a43ceafed2b9abc/Sources/ComposableArchitecture/Effect.swift#L250-L260
- Nesting `merge` and subjects can cause effect to prematurely end: https://github.com/pointfreeco/swift-composable-architecture/issues/946
- Concatenation behavior differs between OS versions: https://github.com/pointfreeco/swift-composable-architecture/blob/dd084dc147dadede0ad25f829a43ceafed2b9abc/Examples/VoiceMemos/VoiceMemosTests/VoiceMemosTests.swift#L11

We even have a dedicated [label](https://github.com/pointfreeco/swift-composable-architecture/issues?q=+label%3A%22apple+bug%22+) tracking bugs reported that are actually in Apple's frameworks.

So my hunch is this is another one 😕

A few potential places to go from here:

- Combine Schedulers' [Timer.swift](https://github.com/pointfreeco/combine-schedulers/blob/84858ad114d8737af5fe58d726d81f15d5b8bd23/Sources/CombineSchedulers/Timer.swift) is mostly copy-pasted from Foundation [here](https://github.com/apple/swift-corelibs-foundation/blob/8a0ed08449db6b7175f7f0c9a677591ff8ef64c8/Darwin/Foundation-swiftoverlay/Publishers%2BTimer.swift), so if the bug _is_ in here we could potentially fix it in Combine Schedulers.
- If this is a more general Combine bug lurking in how we implement an `Effect` API, we're open to adding another workaround if someone tracks down the issue
- If this is a Combine bug in how we implement the case study, we're also open to adding a workaround, though we're less worried about the problem in demo code, especially if it's mostly out of our hands

**Comment by tgrapperon** [comment]
@BeerPoldet I'm not able to reproduce the leaks on my side, Xcode 13.2.1 & 13.3b.
What I'm doing:
```
Launch SwiftUI Case-Studies 
  > Long-living Effects 
  > Navigate to another screen [X] 
  < Long-living Effect [X]
  > Navigate to another screen [X]
  < Long-living Effect [X]
…
```
`[X]` are the moments where I'm snapshotting the memory to look for leaks.
Is there another specific chain of actions that can reproduce relatively reliably the leaks?

**Comment by BeerPoldet** [comment]
@tgrapperon, Thanks for your response. Following your step. these are my result. I don't think I did anything special.
```
Launch SwiftUI Case-Studies 
  > Long-living Effects [0]
  > Navigate to another screen [1]
  < Long-living Effect [2]
  > Navigate to another screen [3]
  < Long-living Effect [4]
  > Navigate to another screen [5]
  < Long-living Effect [6]
  > Navigate to another screen [7]
  < Long-living Effect [8]
```

The number in square brackets indicates when I took the memory graph snapshots.
I have images for each iteration below.

**Graph at [0]**
<img width="1045" alt="Screen Shot 2022-02-06 at 10 30 13" src="https://user-images.githubusercontent.com/7531231/152666820-ccffd0af-dcfa-4a15-8846-94924ff3b032.png">

**Graph at [1] & [2]** (since the result are quite the same, so I added just one image)
<img width="1412" alt="Screen Shot 2022-02-06 at 10 32 10" src="https://user-images.githubusercontent.com/7531231/152666830-50d23e32-0270-4be0-be84-50fc37cc22cf.png">

**Graph at [3] & [4]**
<img width="1461" alt="Screen Shot 2022-02-06 at 10 32 43" src="https://user-images.githubusercontent.com/7531231/152666836-ab4c2f0c-b82d-427a-a243-2a4b7a6fd90a.png">

**Graph at [5] & [6]**
<img width="1468" alt="Screen Shot 2022-02-06 at 10 34 08" src="https://user-images.githubusercontent.com/7531231/152666839-64008053-8246-4ad8-aa70-461d4b23a26e.png">

**Graph at [6] & [8]**
<img width="1470" alt="Screen Shot 2022-02-06 at 10 34 29" src="https://user-images.githubusercontent.com/7531231/152666853-8e4d7916-9eea-4fe3-940e-af973fa6c170.png">

I'm guessing that the leaks came from the fact that I'm using the Apple M1 chip.

**Comment by iampatbrown** [comment]
@BeerPoldet @tgrapperon I have a feeling this is going to be because of #949. I didn't think to check if using `Publishers.PrefixUntilOutput` would cause a leak... I assumed it would't :(...  I can confirm this a bit later, but my guess is there won't be a leak prior to #949.

If this is the case, implementing a custom publisher for cancellations might be the best bet. 

Good catch @BeerPoldet.

**Comment by BeerPoldet** [comment]
@iampatbrown You guessed it right. I've checked out TCA 0.32.0 and tested it. No leaks! 🎉

**Comment by tgrapperon** [comment]
Damn! Good catch @iampatbrown. I was actually using `main` on my fork which was a few commits behind!
Sorry for misleading you @BeerPoldet!

**Comment by iampatbrown** [comment]
@tgrapperon ahh! I've done that before too :P 

Solid work both of you. I can have a crack at fixing this tonight.

**Comment by BeerPoldet** [comment]
Confirm fixed in #997
